datafusion/common/src/scalar.rs

```diff
@@ -2067,30 +2071,78 @@ impl ScalarValue {
     ///
     /// assert_eq!(scalar_vec, expected);
     /// ```
-    pub fn convert_array_to_scalar_vec(array: &dyn Array) -> Result<Vec<Vec<Self>>> {
-        let mut scalars = Vec::with_capacity(array.len());
+    pub fn convert_list_array_to_scalar_vec<O: OffsetSizeTrait>(
```

Comment:
I think this would be easier to use if it didn't mix generic and non generic code

Maybe something like

```suggestion
    pub fn convert_list_array_to_scalar_vec(
      array: &dyn Array,
    ) -> Result<Vec<Vec<Self>>> {    
        if let Some(arr) = array.as_list_opt::<i32> {
            Self::convert_list_array_to_scalar_vec_internal(arr)
        } else if let Some(arr) = array.as_list_opt::<64> {
            Self::convert_list_array_to_scalar_vec_internal(arr)
        } else {
            _internal_err!("Expected GenericListArray but found: {array:?}")
        }
    }
```

And then internally pass in the cast Array  by changing

```
    fn convert_list_array_to_scalar_vec_internal<O: OffsetSizeTrait>(
        array: &dyn Array,
    ) -> Result<Vec<Vec<Self>>> {
```

to

```
    fn convert_list_array_to_scalar_vec_internal<O: OffsetSizeTrait>(
        array: &GenericListArray<O>,
    ) -> Result<Vec<Vec<Self>>> {
```